projects/Mallard/src/helpers/promise-maybe.ts

```diff
@@ -0,0 +1,26 @@
+interface PendingPromise<T> {
+    state: 'unresolved'
+    getValue: () => Promise<T>
+}
+
+interface ResolvedPromise<T> {
+    state: 'resolved'
+    value: T
+}
+
+export type PromiseMaybe<T> = ResolvedPromise<T> | PendingPromise<T>
+
+const isPromise = <T>(promise: PromiseMaybe<T>): promise is PendingPromise<T> =>
+    promise.state === 'unresolved'
```

Comment:
So I think this is important because a `then` will never be called until the current event loop completes- and we want the data now if the promise has resolved. I can't figure out how this would change to `'resolved'` on resolution. I think we almost want some kind of structure that can be subscribed to in a hook.